### PR TITLE
[GLUTEN-1201][CORE] Skip scalar subquery execution during validation

### PR DESCRIPTION
**Pull request by loudongfeng**
## What changes were proposed in this pull request?

Please check #1201 

## How was this patch tested?

Test manually by explain tcpds q14-1, scale factor 0.1

### CODE REVIEW

**Comment by PHILO-HE** [comment]
So this is an optimization to avoid the unnecessary execution of scalar subquery in validation phase. Currently, it does in validation phase, but there is no functionality issue. Right?

**Comment by loudongfeng** [comment]
Yes, only performance issue.

File: gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:167]: Do we need to set validation state also for the other operators?

**Review by rui-mo — COMMENTED** [review]
Looks the purpose of this PR is partially the same as https://github.com/oap-project/gluten/pull/1097.

File: gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala
Comment (loudongfeng) [reply gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:167]: I've thought about making a general mechanism, but didn't find any other scenarios.For example, doExecuteBroadcast for broadCastJoin, rddForSampling in RangePartitioner are not collected during validation.

**Comment by loudongfeng** [comment]
> Looks the purpose of this PR is partially the same as #1097.

#1097 can fix this issue too.

File: gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala
Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:167]: Thanks for your reply. I think we can continue to add it if we find other case. Just want to confirm, is any Subquery containing a Projection?

Comment (loudongfeng) [reply gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:167]: Yes, subquery will appear in many places. Including filter, projection, aggregation expressions, etc.
Here is an example 
```sql
select (select avg(t_time_sk) from time_dim) as avg , (select max(t_time_sk) from time_dim) + (select min(t_time_sk) from time_dim) as range
```
After your reminder, i think i need to make some change in this PR to cover more scenarios.

Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:167]: Thank you for that. I agree we can cover more common scenarios first in this PR.

File: gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala
Comment (loudongfeng) [reply gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:32]: @rui-mo Aggreed. Method in `ExpressionBuilder.makeLiteral` has some validation logic  that rely on actual subquery values.

Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:32]: Do you think it sounds good to add a config to control whether subquery is executed during validation? By default, the subquery validation still takes effect.

Comment (loudongfeng) [reply gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:33]: I plan to skip execution only for data insensitive data Types. For data sensitive dataTypes like StructType, may be we can add a result cache or something similar to that(the idea is immature).

Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:33]: This is OK to me if it can bypass this issue. Can we move the `valueSensitiveDataType` method into ScalarSubqueryTransformer? Because looks it helps handle a special case in this case, and is not generally used.

**Review by rui-mo — DISMISSED** [review]

**Review by zzcclp — DISMISSED** [review]
LGTM

File: gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala
Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:34]: > When scala subquery filters pushed down to scan by gluten, duplicate execution happens too. It's not in validation stage.

@loudongfeng Could you explain more about why duplication happens?

Comment (loudongfeng) [reply gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:34]: This query can reproduce:
```sql
explain select i_item_sk from item where i_product_name = (select max(i_product_name) from item);
```
SparkSQL output
```sql
== Physical Plan ==
*(1) Project [i_item_sk#7L]
+- *(1) Filter (isnotnull(i_product_name#28) AND (i_product_name#28 = Subquery scalar-subquery#1, [id=#38]))
   :  +- Subquery scalar-subquery#1, [id=#38]
   :     +- SortAggregate(key=[], functions=[max(i_product_name#52)])
   :        +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=36]
   :           +- SortAggregate(key=[], functions=[partial_max(i_product_name#52)])
   :              +- *(1) ColumnarToRow
   :                 +- FileScan parquet tpcds_parquet.item[i_product_name#52] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/nemon/tpcd..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i_product_name:string>
   +- *(1) ColumnarToRow
      +- FileScan parquet tpcds_parquet.item[i_item_sk#7L,i_product_name#28] Batched: true, DataFilters: [isnotnull(i_product_name#28)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/nemon/tpcd..., PartitionFilters: [], PushedFilters: [IsNotNull(i_product_name)], ReadSchema: struct<i_item_sk:bigint,i_product_name:string>

```
Gluten output
```sql
== Physical Plan ==
CHNativeColumnarToRow
+- ^(2) ProjectExecTransformer [i_item_sk#7L]
   +- ^(2) FilterExecTransformer (isnotnull(i_product_name#28) AND (i_product_name#28 = Subquery scalar-subquery#1, [id=#57]))
      :  +- Subquery scalar-subquery#1, [id=#57]
      :     +- SortAggregate(key=[], functions=[max(i_product_name#52)])
      :        +- CHNativeColumnarToRow
      :           +- CoalesceBatches
      :              +- ColumnarExchange SinglePartition, ENSURE_REQUIREMENTS, false, [plan_id=53], [id=#53], [OUTPUT] List(max:StringType)
      :                 +- RowToCHNativeColumnar
      :                    +- SortAggregate(key=[], functions=[partial_max(i_product_name#52)])
      :                       +- CHNativeColumnarToRow
      :                          +- ^(1) NativeFileScan parquet tpcds_parquet.item[i_product_name#52] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/nemon/tpcd..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i_product_name:string>
      +- ^(2) NativeFileScan parquet tpcds_parquet.item[i_item_sk#7L,i_product_name#28] Batched: true, DataFilters: [isnotnull(i_product_name#28), (i_product_name#28 = Subquery scalar-subquery#1, [id=#57])], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/nemon/tpcd..., PartitionFilters: [], PushedFilters: [IsNotNull(i_product_name)], ReadSchema: struct<i_item_sk:bigint,i_product_name:string>
            +- Subquery scalar-subquery#1, [id=#57]
               +- SortAggregate(key=[], functions=[max(i_product_name#52)])
                  +- CHNativeColumnarToRow
                     +- CoalesceBatches
                        +- ColumnarExchange SinglePartition, ENSURE_REQUIREMENTS, false, [plan_id=53], [id=#53], [OUTPUT] List(max:StringType)
                           +- RowToCHNativeColumnar
                              +- SortAggregate(key=[], functions=[partial_max(i_product_name#52)])
                                 +- CHNativeColumnarToRow
                                    +- ^(1) NativeFileScan parquet tpcds_parquet.item[i_product_name#52] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/nemon/tpcd..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i_product_name:string>

```
In Gluten plan, scalar-subquery#1 is pushed down to scan,so `ScalarSubqueryTransformer.doTransform` execute twice during whole stage transform.One for FilterExecTransformer, one for TableScanTransformer.
May be we can fix this in another PR.

Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:34]: Thank you. Yes, we can do that.
There are some logics in Filter transformer to prevent duplicate execution for the filter conditions pushed down into Scan. Do you suggest although with this logic the Subquery is still executed twice?

https://github.com/oap-project/gluten/blob/main/gluten-data/src/main/scala/io/glutenproject/execution/GlutenFilterExecTransformer.scala#L39-L44

Comment (loudongfeng) [reply gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:34]: Thanks for your information. I find that this duplicate execution only happens for ch backend + parquet file format. After changing this line from FilterExecTransformer to CHFilterExecTransformer, then no duplicate execution.
https://github.com/oap-project/gluten/blob/main/backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHSparkPlanExecApi.scala#L92

@zzcclp Any idea why we use two different FilterExecTransformer for ch backend?

Comment (loudongfeng) [reply gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:34]: I have created a seperate ticket to track the issue mentioned above. #1227 
Shall we merge this PR first? @rui-mo

File: gluten-core/src/main/scala/io/glutenproject/expression/TransformerState.scala
Comment (PHILO-HE) [inline gluten-core/src/main/scala/io/glutenproject/expression/TransformerState.scala:1]: Please add license header for this source file.

**Review by PHILO-HE — APPROVED** [review]

File: gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala
Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:34]: Sure. If there's no other comment.

**Comment by PHILO-HE** [comment]
@Wangguangxin

**Comment by WangGuangxin** [comment]
@PHILO-HE The change LGTM to avoid execute subquery when do validation. 

But I think there is another improvement that's submit subquery concurrently before `doTransform`, by utilize `SparkPlan.prepare()`, like https://github.com/oap-project/gluten/pull/1097/files#diff-2229bea3ccd763cfbccc11ad3337540c895b216b5568a10b1ddaee56919862faR184

What do you think?

**Comment by PHILO-HE** [comment]
> @PHILO-HE The change LGTM to avoid execute subquery when do validation.
> 
> But I think there is another improvement that's submit subquery concurrently before `doTransform`, by utilize `SparkPlan.prepare()`, like https://github.com/oap-project/gluten/pull/1097/files#diff-2229bea3ccd763cfbccc11ad3337540c895b216b5568a10b1ddaee56919862faR184
> 
> What do you think?

I agree. Please keep that part of fix in your PR and rebase your code after this patch is merged.

**Comment by PHILO-HE** [comment]
Will merge the patch soon. Thanks for the contribution!

**Comment by loudongfeng** [comment]
Many thanks for your review. @rui-mo @PHILO-HE @zzcclp @WangGuangxin

File: gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala
Comment (zhztheplayer) [inline gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala:61]: Hi @loudongfeng , I just bumped into this code. Do you still remember the reason of using this condition?

**Pull request merged**
